File: NimdaConBackEnd/backend/src/auth/auth.controller.ts
import { Controller, Post, Body, UnauthorizedException, HttpStatus, HttpCode } from '@nestjs/common';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { RegisterDto } from './dto/register.dto';

@Controller('api/auth')
export class AuthController {

  constructor(private authService: AuthService) {}

  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(@Body() loginDto: LoginDto) {
    try {
      // #1 사용자 검증 로직 
      const user = await this.authService.validateUser(
        loginDto.username, // 유저가 입력한 사용자명
        loginDto.password, // 유저가 입력한 비밀번호 
      );
      
      // #1.1 사용자 검증 실패 시 예외 처리 
      if (!user) {
        throw new UnauthorizedException('Invalid credentials');
      }
      
      // #1.2 사용자 검증 성공 시 로그인 처리 시만
      const result = await this.authService.login(user);
      return {
        success: true,
        message: 'Login successful',
        data: result
      };
    } catch (error) {
      return {
        success: false,
        message: error.message || 'Login failed',
        data: null
      };
    }
  }

  // #2 회원가입 로직 
  @Post('register')
  @HttpCode(HttpStatus.CREATED)
  async register(@Body() registerDto: RegisterDto) {
    try {
      // #2.1 회원가입 처리
      const user = await this.authService.register(
        registerDto.username, 
        registerDto.password, 
        registerDto.email
      );
      
      // #2.2 회원가입 성공 시 JWT 토큰 발급 (자동 로그인)
      const result = await this.authService.login(user);
      return {
        success: true,
        message: 'Registration successful',
        data: result
      };
    } catch (error) {
      return {
        success: false,
        message: error.message || 'Registration failed',
        data: null
      };
    }
  }
}

/**
 * @Body() - Request에 포함된 데이터를 받아온다. 
 * 여기서는 사용자가 로그인 폼에 입력한 데이터 {username: string, password: string} 문자열 검증 
 */

File: NimdaConBackEnd/backend-spring/src/main/java/com/nimda/con/user/controller/AuthController.java
package com.nimda.con.user.controller;

import com.nimda.con.user.dto.LoginDTO;
import com.nimda.con.user.dto.RegisterDTO;
import com.nimda.con.user.entity.User;
import com.nimda.con.user.service.AuthService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/auth")
@CrossOrigin(origins = "*")
public class AuthController {
    
    @Autowired
    private AuthService authService;
    
    /**
     * 로그인
     * @param loginRequest 로그인 요청 데이터
     * @return JWT 토큰과 사용자 정보
     */
    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginDTO loginRequest) {
        try {
            // 사용자 인증
            Optional<User> userOpt = authService.validateUser(
                loginRequest.getUsername(), 
                loginRequest.getPassword()
            );
            
            if (userOpt.isPresent()) {
                // 로그인 성공
                Map<String, Object> response = authService.login(userOpt.get());
                return ResponseEntity.ok(response);
            } else {
                // 인증 실패
                Map<String, String> error = new HashMap<>();
                error.put("message", "Invalid username or password");
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(error);
            }
        } catch (Exception e) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "Login failed: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error);
        }
    }
    
    /**
     * 회원가입
     * Request Data : Register DTO (username, password, email)
     */
    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterDTO registerRequest) {

        try {

            User user = authService.register(
                registerRequest.getUsername(),
                registerRequest.getPassword(),
                registerRequest.getEmail()
            );
            
            return ResponseEntity.status(HttpStatus.CREATED).body(user);
        } 
        
        catch (RuntimeException e) {
            Map<String, String> error = new HashMap<>();
            error.put("message", e.getMessage());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
        } 
        
        catch (Exception e) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "Registration failed: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error);
        }

    }

}
